### PR TITLE
[MOB-4754] add lastPushPayload to keychain

### PR DESCRIPTION
**Pull request by roninopf**
## 🔹 Jira Ticket(s)

* [MOB-4754](https://iterable.atlassian.net/browse/MOB-4754)

## ✏️ Description

* this makes `lastPushPayload` store in `IterableKeychain` rather than `IterableUserDefaults` to save it securely

### CODE REVIEW

**Review by davidtruong — APPROVED** [review]
LFTM

File: swift-sdk/Internal/LocalStorage.swift [hunk 1]
```diff
@@ -117,4 +122,12 @@ struct LocalStorage: LocalStorageProtocol {
...
         }
     }
+    
+    private func moveLastPushPayloadFromUserDefaultsToKeychain() {
+        if let (userDefaultLastPushPayload, expiration) = iterableUserDefaults.getLastPushPayloadAndExpirationPair() {
+            keychain.setLastPushPayload(userDefaultLastPushPayload, withExpiration: expiration)
+            
+            ITBInfo("UPDATED: moved lastPushPayload from UserDefaults to IterableKeychain")
+        }
+    }
```
Comment (evangreer91) [inline swift-sdk/Internal/LocalStorage.swift:132]: LGTM

File: swift-sdk/Internal/LocalStorageProtocol.swift [hunk 2]
```diff
@@ -6,16 +6,27 @@ import Foundation
...
     func getAttributionInfo(currentDate: Date) -> IterableAttributionInfo?
+    
     func save(attributionInfo: IterableAttributionInfo?, withExpiration expiration: Date?)
-    func getPayload(currentDate: Date) -> [AnyHashable: Any]?
-    func save(payload: [AnyHashable: Any]?, withExpiration: Date?)
+    
+    func getLastPushPayload(_ currentDate: Date) -> [AnyHashable: Any]?
+    
+    func saveLastPushPayload(_ payload: [AnyHashable: Any]?, withExpiration expiration: Date?)
+
```
Comment (evangreer91) [inline swift-sdk/Internal/LocalStorageProtocol.swift:29]: Looks good. Aligns with LocalStorage.swift.

File: tests/common/CommonMocks.swift [hunk 3]
```diff
@@ -465,18 +466,11 @@ class MockLocalStorage: LocalStorageProtocol {
...
-                // expired
-                return true
-            }
-        } else {
-            // no expiration
+        guard let expiration = expiration else {
             return false
         }
+        
+        return !(expiration.timeIntervalSinceReferenceDate > currentDate.timeIntervalSinceReferenceDate)
```
Comment (evangreer91) [inline tests/common/CommonMocks.swift:473]: Nice refactor. Follows what you did above.

File: tests/unit-tests/LocalStorageTests.swift [hunk 4]
```diff
@@ -96,7 +111,8 @@ class LocalStorageTests: XCTestCase {
     
     func testPayload() throws {
         let mockDateProvider = MockDateProvider()
-        let localStorage = LocalStorage(userDefaults: LocalStorageTests.getTestUserDefaults())
+        let localStorage = MockLocalStorage()
+
```
Comment (evangreer91) [inline tests/unit-tests/LocalStorageTests.swift:115]: extra line here.

File: swift-sdk/Internal/KeychainWrapper.swift [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    private let SecAttrService: String = kSecAttrService as String
+    private let SecAttrGeneric: String = kSecAttrGeneric as String
+    private let SecAttrAccount: String = kSecAttrAccount as String
+    private let SecAttrSynchronizable: String = kSecAttrSynchronizable as String
+    private let CFBooleanTrue = kCFBooleanTrue
+    private let CFBooleanFalse = kCFBooleanFalse
+    private let SecMatchLimit: String = kSecMatchLimit as String
+    private let SecMatchLimitOne = kSecMatchLimitOne
+    private let SecReturnData: String = kSecReturnData as String
+}
```
Comment (evangreer91) [inline swift-sdk/Internal/KeychainWrapper.swift:117]: Looks good to me. Sets up the methods to interact directly with the keychain.

**Review by evangreer91 — COMMENTED** [review]
Looks good to me. I added a few minor questions and minor changes.

File: swift-sdk/Internal/InternalIterableAPI.swift [hunk 6]
```diff
@@ -522,7 +522,7 @@ final class InternalIterableAPI: NSObject, PushTrackerProtocol, AuthProvider {
         let expiration = Calendar.current.date(byAdding: .hour,
                                                value: Const.UserDefault.payloadExpiration,
                                                to: dateProvider.currentDate)
-        localStorage.save(payload: payload, withExpiration: expiration)
+        localStorage.saveLastPushPayload(payload, withExpiration: expiration)
```
[the remark this answers is not in the document]
Comment (roninopf) [reply swift-sdk/Internal/InternalIterableAPI.swift:525]: Yep.

**Pull request merged**